include/SDL3/SDL_pen.h

```diff
@@ -0,0 +1,265 @@
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+
+/**
+ *  \file SDL_pen.h
+ *
+ *  Include file for SDL pen event handling.
+ *
+ *  This file describes operations for pressure-sensitive pen (stylus and/or eraser) handling, e.g., for input
+ *    and drawing tablets or suitably equipped mobile / tablet devices.
+ *
+ *  To get started with pens:
+ *  - Listen to ::SDL_PenMotionEvent and ::SDL_PenButtonEvent
+ *  - To avoid treating pen events as mouse events, ignore  ::SDL_MouseMotionEvent and ::SDL_MouseButtonEvent
+ *    whenever "which" == ::SDL_PEN_MOUSEID.
+ *
+ *  This header file describes advanced functionality that can be useful for managing user configuration
+ *    and understanding the capabilities of the attached pens.
+ *
+ *  We primarily identify pens by ::SDL_PenID.  The implementation makes a best effort to relate each :SDL_PenID
+ *    to the same physical device during a session.  Formerly valid ::SDL_PenID values remain valid
+ *    even if a device disappears.
+ *
+ *  For identifying pens across sessions, the API provides the type ::SDL_GUID .
+ */
+
+#ifndef SDL_pen_h_
+#define SDL_pen_h_
+
+#include "SDL_error.h"
+#include "SDL_guid.h"
+#include "SDL_stdinc.h"
+
+/* Set up for C function definitions, even when using C++ */
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef Uint32 SDL_PenID; /**< SDL_PenIDs identify pens uniquely within a session */
+
+#define SDL_PEN_INVALID ((Uint32)0) /**< Reserved invalid ::SDL_PenID is valid */
+
+#define SDL_PEN_MOUSEID ((Uint32)-2) /**< Device ID for mouse events triggered by pen events */
+
+#define SDL_PEN_INFO_UNKNOWN (-1) /**< Marks unknown information when querying the pen */
+
+/**
+ * Pen axis indices
+ *
+ * Below are the valid indices to the "axis" array from ::SDL_PenMotionEvent and ::SDL_PenButtonEvent.
+ * The axis indices form a contiguous range of ints from 0 to ::SDL_PEN_AXIS_LAST, inclusive.
+ * All "axis[]" entries are normalised to either 0..1 (unidirectional axes) or between -1..1 (bidirectional axes).
+ * Unsupported entries are always "0.0f".
+ *
+ */
+typedef enum
+{
+    SDL_PEN_AXIS_PRESSURE = 0,               /**< Pen pressure.  Unidirectional: 0..1.0 */
+    SDL_PEN_AXIS_XTILT,                      /**< Pen horizontal tilt fraction.  Bidirectional: -1.0..1.0 (left-to-right).  This is NOT the angle, but the tilt vector x component.  The physical max/min tilt may be smaller than -1.0 / 1.0, cf. \link SDL_PenCapabilityInfo \endlink .  */
+    SDL_PEN_AXIS_YTILT,                      /**< Pen vertical tilt fraction.  Bidirectional: -1.0..1.0 (top-to-bottom).  This is NOT the angle, but the tilt vector y component.  */
```

Comment:
Most APIs seem to report tilt angles instead, either relative to the rotation angle (like [iPadOS](https://developer.apple.com/documentation/uikit/touches_presses_and_gestures/illustrating_the_force_altitude_and_azimuth_properties_of_touch_input#3138986)) or relative to the Z axis ([Win32](https://learn.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-pointer_pen_info) and [Wayland](https://wayland.app/protocols/tablet-unstable-v2#zwp_tablet_tool_v2:event:tilt)). If you're going to report tilt vectors, it would be useful to detail how users can convert these into representations they're more familiar with.

Reply:
The only reason you'd want a vector here is to indicate magnitude, this should use an angle, defined relative to the Z axis.

Reply:
I find vectors simpler to understand and more useful downstream-- I can't think of a practical application right now where I would do anything with tilt angles other than to put them into trigeonometric functions to get a 2D vector that represents the direction and magnitude of the pen tilt.  The other concern I have about reporting angles is that this seems more likely to contribute to subtle bugs, since there are two different, meaningful ways to interpret two angles (as @cgutman noted) that produce very similar results when the rotation/horizontal tilt angle is small.  A smaller  point in favour is that reporting vectors is also more in line with joystick and gamepad axis events (we can think of the x-tilt and the y-tilt as two separate axes).

XInput2 seems to report tilt vectors, too, so it's not unprecedented-- and my (limited) understanding of the market is that  XInput2 is industrially relevant, due to tools like [Maya](https://help.autodesk.com/view/MAYAUL/2023/ENU/) that have been ported around since the IRIX days.

I'd prefer to improve the documentation instead, but please do let me know if I am missing something or seem to be completely off the mark here...

Reply:
fwiw, XInput2 (and 1.x) basically blindly forwards whatever the device/kernel gives it, there's no logic anywhere to even understand that something is a vector vs an angle. For Wacom tablets those values are degrees.

libinput exposes those as degrees too which is where the Wayland support for degrees comes from.

Reply:
@whot, thanks for correcting me!

I ran some experiments to convince myself, and my Cintiq 16 reports (roughly) the same values on Wayland as on XInput2, after normalisation (so  right now those two backends for this patch don't agree about what to return).
I'll take your word on it that these are angles; this turns out to be a bit tricky to check just by looking at it, since the differences between $\frac{t}{\pi}$ and $\sin(t)$ is at most around 14 degrees.
I think I'll update the `testpen` tool to make it easier to check this.

(One odd bit: the angles reported by XInput2 and Wayland seem a bit off-- Wayland never seems to report an angle of 0 on my setup, while XInput2 does.  I'll need to double-check this.)

Reply:
@creichen Which driver are you using under xorg? the libinput or the wacom driver/ (check with `xinput list-props <device name>`, the prefix will tell you. If it's the libinput driver the values should be the same as Wayland since it just forwards the data from libinput. If it's the wacom driver that may explain minor differences though I'd have to check why they occur.

Wayland will certainly suffer from rounding errors due to the 24.8 fixed point limitation though.

`libinput debug-events` will tell you what comes out of libinput btw.

Reply:
I'm not aware of any API that returns a vector like this. I'm not saying it is a poor decision, but it *is* a unique one.

> I find vectors simpler to understand and more useful downstream-- I can't think of a practical application right now where I would do anything with tilt angles other than to put them into trigeonometric functions to get a 2D vector that represents the direction and magnitude of the pen tilt.

I agree that the angles themselves often need some degree of processing through trig functions to be useful, but so does this vector. Most applications that I'm aware of are primarily interested in having some UI element rotate to match the azimuth angle of the pen. I can't use the vector as-is in a 2D rotation matrix; I have to feed it through atan2() first (and *then* through sin() and cos() just as if I had been given the angle directly). The next most common thing I've seen is to adjust brush properties (e.g. shape or texture or color) in response to the altitude angle of the pen. I again can't directly use the vector representation as-is here; I'd need to calculate its magnitude at minimum (for properties that are linear in the angle's projection) or possibly its arcsin() (for properties that are linear in the angle itself).

>  The other concern I have about reporting angles is that this seems more likely to contribute to subtle bugs, since there are two different, meaningful ways to interpret two angles (as @cgutman noted) that produce very similar results when the rotation/horizontal tilt angle is small.

I don't see how a vector is much of an improvement here. This representation _also_ produces similar results at small angles --- especially to a representation using tilt-x and tilt-y in radians.

Reply:
I also vote for angle, as it's easier for me to wrap my head around, and is more likely to translate into something I can directly use.

Reply:
Fair points; let's go for angles, then.  I'll write a quick summary below.